src/tallies/MeshTally.C

```diff
@@ -82,6 +98,30 @@ MeshTally::MeshTally(const InputParameters & parameters)
     if (isParamValid("mesh_translation"))
       mooseError("The mesh filter cannot be translated if directly tallying on the mesh "
                  "provided in the [Mesh] block!");
+
+    // Fetch subdomain IDs for block restrictions.
+    if (isParamValid("blocks"))
+    {
+      auto block_names = getParam<std::vector<SubdomainName>>("blocks");
+      if (block_names.empty())
+        mooseError("Subdomain names must be provided if using 'blocks'!");
+
+      auto block_ids = _mesh.getSubdomainIDs(block_names);
+      std::copy(
+          block_ids.begin(), block_ids.end(), std::inserter(_tally_blocks, _tally_blocks.end()));
+
+      // Check to make sure all of the blocks are in the mesh.
```

Comment:
When restricting tallying only to certain blocks, do the results still get mapped back to the [Mesh] properly? Originally, we would just map element 4 to element 4, etc. Do your tests cover the scenario where the blocks we want to tally are not the "first" in the mesh?

Reply:
Both test cases added in this PR covering block restriction check this. We block restrict such that tallying doesn't happen on the first pebble (which is the first set of elements in the mesh), and so the elements in the created submesh need to get mapped back to the full mesh. To confirm this is happening correctly I compared the results of these tests to an equivalent where the mesh mirror only contains two pebbles.

Reply:
perfect, thanks!